Move compressed marshaling away from the stdlib and into the compilers (5.1 version)

xavierleroy opened the pull request:
This is the second attempt at fixing the problem reported in #12562 concerning the compressed marshaling facility introduced in OCaml 5.1 by #12006.

The first attempt #12705 kept the API introduced in #12006 (an extra flag `Marshal.Compression` in the `Marshal` standard library module) and tried to make sure that the ZSTD library is dynamically linked with ocamlopt-generated executables only if the `Marshal` module is actually used.  This failed because of limitations of some system linkers.

The present attempt removes the `Marshal.Compression` API and makes compressed marshaling unavailable via the standard library and private to the OCaml compilers.  A new module `Compression` is added to the compilers and to compiler-libs, and used to compress and uncompress the same compilation artefacts as in #12006.  (Unless ZSTD is not available or configured out, in which case plain marshaling is used.)

The implementation reuses parts of the first attempt #12705:
- The ZSTD-specific C code is moved to runtime/zstd.c and hooks into extern.c and intern.c when activated.
- The bytecode interpreter `ocamlrun` and executables produced with `ocamlc -custom` always include zstd.c and always dynamically link with ZSTD.  (Doing otherwise complicates coldstart and bootstrap too much.)

What's new in this second attempt:
- Native-code executables produced by `ocamlopt` do NOT include zstd.c and do NOT link with ZSTD, unless they use compiler-libs (see next).
- zstd.c and -lzstd are automatically included when linking with ocamlcommon.cmxa, the first part of native-code compiler-libs.  This makes them usable from ocamlc.opt, ocamlopt.opt, and any native-code program that uses compiler-libs and therefore needs to work with compressed compilation artifacts.

This PR is against 5.1, so that it can be integrated in release 5.1.1.  Two points of attention:
- Accepting this PR means that we introduce an API incompatibility between 5.1.1 and 5.1 (the removal of the `Marshal.Compression` flag).  IIRC, we never did this (a minor release with an incompatible API) before.  The working group composed of @dra27, @octachron, @nojb and I think this is acceptable given the circumstances.  This will have to be confirmed at the next developers meeting.
- The implementation of this PR was intentionally kept small and a bit hackish to minimize the changes in 5.1.1.  If it is accepted, we (the working group) have ideas to clean up the implementation and perhaps make compressed marshaling accessible to users again (as a separate library in otherlibs/, explicitly linked, like the Unix library) in time for 5.2.

xavierleroy committed:
Reduce the dependencies of native-code Dynlink

Currently, both the bytecode version and the native-code version of
Dynlink drag in many modules from the compiler, including some that
use compressed marshaling.  This causes users of dynlink.cmxa to
depend on -lzstd systematically.

Actually, the native-code version of Dynlink needs much fewer compiler
modules, none of which require compressed marshaling.

This commit just shrinks the list of compiler modules included in
dynlink.cmxa, thus cutting the dependency on -lzstd.

xavierleroy committed:
Move the functions that use ZSTD compression to a separate zstd.c file

These functions used to be in extern.c and intern.c.
Now, they are in zstd.c and connect themselves to extern.c and intern.c
via hooks activated by a new primitive `caml_zstd_initialize`.

dra27 committed:
Remove Marshal.Compression and Marshal.compression_supported

- Removes the Marshal.Compression constructor, in preparation for
  explicit API in 5.2
- Add just to_channel (for now) to ocamlcommon
- caml_zstd_initialize called from Compression for now.

xavierleroy committed:
Avoid dependency from Cmxs_format to Misc

Already done in trunk.

xavierleroy committed:
Internal API change: `Compression.to_channel` -> `Compression.output_value`

The "flags" argument was always [], so remove it.
This avoids an Obj.magic in compression.ml.

xavierleroy committed:
Correct libs used in output-complete-obj test

Manual cherry-pick from 7ed72683b8 .

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

xavierleroy committed:
ocamlopt: do not link -lzstd nor zstd.n.o by default, only for compilerlibs

We attach -lzstd to ocamlcommon.cmxa and insert zstd.n.o in ocamlcommon.a.
(The latter is a hack, to be replaced by something better eventually.)

xavierleroy committed:
Internal API change: introduce and use `Compression.input_value`

Under the hood, it's just `Stdlib.input_value`.
However, it documents the places in the compiler where we expect to read
compressed marshaled data.

xavierleroy committed:
Bootstrap after removal of Marshal.Compression flag

xavierleroy commented on compilerlibs/Makefile.compilerlibs:
```diff
@@ -430,8 +432,15 @@ compilerlibs/ocamlcommon.cma: $(COMMON_CMI) $(ALL_CONFIG_CMO) $(COMMON)
+compilerlibs/ocamlcommon.cmxa: $(COMMON_CMI) $(COMMON:.cmo=.cmx) $(ZSTD_OBJ)
+	$(V_LINKOPT)$(CAMLOPT) -a -linkall -o $@ \
+                               $(ZSTD_FLAGS) $(COMMON:.cmo=.cmx)
+	$(V_MKLIB)$(ARCMD) r compilerlibs/ocamlcommon.$(A) $(ZSTD_OBJ)
```
This creative use of `ar` is due to @dra27.  If we just add `zstd.n.o` to the `ocamlopt -a` command line, this file name is recorded in `ocamlcommon.cmxa` and passed to the system linker when the latter file is used, but the system linker will not find it and will not search for it in the `-L` search path.  Instead, we include `zstd.n.o` and its contents in the `ocamlcommon.a` archive, where the system linker will find it reliably.

It's unclear whether this hack can work with the MSVC librarian `lib`.  Good thing we're not supporting MSVC again yet.

To make things cleaner in the future, we can either
- Change `ocamlopt -a` so that all `.o` files given on the command line are included in the generated `.a` file.  The current behavior (of recording the file names in the `.cmxa` file) is useless anyway.
- Put `zstd.n.o` in a proper `.a` archive, and install this archive.  This would be the case if we make an `otherlibs` that exposes compressed marshaling to users.

kit-ty-kate reviewed (approved):
Successfully tested on macOS, FreeBSD and Alpine Linux. Thank you so much!

xavierleroy commented:
The incompatible API change (removal of `Marshal.Compression` flag and `Marshal.compression_available` function) was approved at this week's developer meeting.  @damiendoligez promised a review of this PR.

damiendoligez commented on configure.ac:
```diff
@@ -1994,7 +1997,6 @@ AS_IF([test x"$with_zstd" != "xno"],
 
 AS_IF([test x"$zstd_status" = "xok"],
   [AC_MSG_NOTICE([compressed marshaling supported])
```
Do we still want to make this notice visible to the user? Do we want to keep the same wording even though compressed marshalling is not available to the user?

xavierleroy replied:
Well spotted.  Let's report "compressed compilation artefacts supported" instead.

damiendoligez commented on otherlibs/dynlink/byte/dynlink.ml:
```diff
@@ -134,7 +134,7 @@ module Bytecode = struct
           else begin
             seek_in ic compunit.cu_debug;
-            [| input_value ic |]
+            [| Compression.input_value ic |]
```
```suggestion
            [| Compression.input_value ic : Instruct.debug_event list array |]
```
I think this is a good opportunity to add the missing type constraint.

xavierleroy replied:
PS.  The correct form was
```
[| (Compression.input_value ic : Instruct.debug_event list) |]
```

damiendoligez reviewed (approved):
Looks good modulo two nitpicks.

damiendoligez commented:
Because #12006 was released, you'll also need a Changes entry for this PR.

xavierleroy committed:
Address review comments on #12734

- Add type constraint in dynlink/byte/dynlink.ml
- Reword configure message about ZSTD support

Co-authored-by: Damien Doligez <damien.doligez@gmail.com>

xavierleroy committed:
Changes for 12734

xavierleroy commented:
CI is green (here and on Jenkins).  Merging in 5.1 branch.  Will port to trunk and improve soon.

hhugo commented on runtime/intern.c:
```diff
@@ -792,32 +789,45 @@ static void caml_parse_header(struct caml_intern_state* s,
+                                       const unsigned char *,
+                                       uintnat) = NULL;
+
+CAMLprim value caml_compression_available(value vunit)
```
This is not used, is it intended ?

Octachron replied:
Yes, indeed. To avoid an unnecessary bootstrap, the primitive will be removed before at release.

hhugo commented on runtime/zstd.c:
```diff
@@ -0,0 +1,115 @@
+
+CAMLprim value caml_zstd_initialize(value vunit)
+{
+  return Val_unit;
```
`return Val_false;`

hhugo commented on runtime/zstd.c:
```diff
@@ -0,0 +1,115 @@
+{
+  caml_extern_compress_output = caml_zstd_compress;
+  caml_intern_decompress_input = caml_zstd_decompress;
+  return Val_unit;
```
`return Val_true;`

Octachron replied:
Fixed in 8abed5e546c6381777ac4036fcdef5139d04590b .
